cmd/main.go

```diff
@@ -380,53 +438,106 @@ func scanEvoChain(ctx context.Context, c *config.Config, client scan.EthClient,
 				waitBeforeNextScan(ctx, c.WaitingTime)
 				break
 			}
-			events, lastScannedBlock, err := s.ScanEvents(ctx, big.NewInt(int64(startingBlock)), big.NewInt(int64(lastBlock)), nil)
-			if err != nil {
-				slog.Error("error occurred while scanning LaosEvolution events", "err", err.Error())
-				break
-			}
 
-			nextStartingBlock, err := storeMintEventsAndUpdateBlock(ctx, stateService, events, lastScannedBlock, client)
+			err = processEvoBlockRange(ctx, client, stateService, s, startingBlock, lastBlock)
 			if err != nil {
+				if errors.Is(err, ReorgError{}) {
+					return err
+				}
 				break
 			}
 
-			startingBlock = nextStartingBlock
+			startingBlock = lastBlock + 1
 		}
 	}
 }
 
-func storeMintEventsAndUpdateBlock(ctx context.Context, stateService state.Service, events []scan.Event, lastBlock *big.Int, client scan.EthClient) (uint64, error) {
+func processEvoBlockRange(ctx context.Context, client scan.EthClient, stateService state.Service, s scan.Scanner, startingBlock, lastBlock uint64) error {
 	tx := stateService.NewTransaction()
 	defer tx.Discard()
+
+	// retrieve the hash of the final block of the previous iteration.
+	prevLastBlockHash, err := tx.GetEvoEndRangeBlockHash()
+	if err != nil {
+		slog.Error("error occurred while reading LaosEvolution end range block hash", "err", err.Error())
+		return nil
+	}
+
+	// During the initial iteration, no hash is stored in the database, so this code block is bypassed.
+	// Verify whether the hash of the last block from the previous iteration remains unchanged; if it differs,
+	// it indicates a reorganization has taken place.
+	if prevLastBlockHash != (common.Hash{}) {
+		var prevIterLastBlock *types.Block
+		prevIterLastBlockNumber := startingBlock - 1
+		prevIterLastBlock, err = client.BlockByNumber(ctx, big.NewInt(int64(prevIterLastBlockNumber)))
+		if err != nil {
+			slog.Error("error occurred while fetching new LaosEvolution start range block", "err", err.Error())
+			return nil
+		}
+
+		if prevIterLastBlock.Hash().Cmp(prevLastBlockHash) != 0 {
+			slog.Error("evolution chain reorganization detected", "block number", startingBlock-1, "chain hash", prevIterLastBlock.Hash().String(), "storage hash", prevLastBlockHash.String())
```

Comment:
I think we should add further instructions after this message, like: "please, wipe out the database and sync again from the desired block number"